### PR TITLE
Add Auth0 user authentication

### PR DESCRIPTION
**Pull request by jancona**
I think this is getting close to ready to merge. The main thing missing is tests of the authentication code itself. I'll take a look at that tomorrow.

### CODE REVIEW

**Comment by DallanQ** [comment]
Wow, you've made a lot of changes here! :-)
I'll look at it first thing in the morning.

File: server/server.go [hunk 1]
```diff
@@ -177,30 +206,21 @@ type Env struct {
...
-		Region:              os.Getenv("AWS_REGION"),
-		BlobStoreEndpoint:   os.Getenv("BLOB_STORE_ENDPOINT"),
-		BlobStoreAccessKey:  os.Getenv("BLOB_STORE_ACCESS_KEY"),
-		BlobStoreSecretKey:  os.Getenv("BLOB_STORE_SECRET_KEY"),
-		BlobStoreBucket:     os.Getenv("BLOB_STORE_BUCKET"),
-		BlobStoreDisableSSL: strings.HasPrefix(strings.ToUpper(os.Getenv("BLOB_STORE_DISABLE_SSL")), "T"),
-		PubSubProtocol:      os.Getenv("PUB_SUB_PROTOCOL"),
-		PubSubPrefix:        os.Getenv("PUB_SUB_PREFIX"),
+	var config Env
+	if err := env.Set(&config); err != nil {
```
Comment (DallanQ) [inline server/server.go:216]: very elegant!

**Comment by DallanQ** [comment]
Finally reviewed this. I'm sorry for taking so long. I don't know where the time went today.

File: server/server.go [hunk 2]
```diff
@@ -152,18 +167,32 @@ func main() {
...
 		// 	Handler(http.StripPrefix("/wasm", http.FileServer(http.Dir(vectyDir))))
-		log.Fatal(http.ListenAndServe(fmt.Sprintf(":%s", env.BaseURL.Port()), // changed from Host because host doesn't work inside docker
-			handlers.LoggingHandler(os.Stdout,
-				handlers.CORS(
-					handlers.AllowedHeaders([]string{"X-Requested-With", "Content-Type", "Authorization"}),
-					handlers.AllowedMethods([]string{"GET", "POST", "PUT", "DELETE", "HEAD", "OPTIONS"}),
-					handlers.AllowedOrigins([]string{"*"}))(r))))
+		if env.BaseURL.Scheme == "https" {
+			log.Fatal(http.ListenAndServeTLS(fmt.Sprintf(":%s", env.BaseURL.Port()),
+				"server.crt", "server.key",
```
[the remark this answers is not in the document]
Comment (jancona) [reply server/server.go:172]: Yes, although I haven't tested this lately. Some of my early auth testing seemed to require `https` endpoints. I can remove it, we could make it the default, or just leave it as an option. If we want it as a default, I can try to switch the Docker/Tilt stuff over.

File: server/server.go [hunk 3]
```diff
@@ -63,13 +71,21 @@ func main() {
...
 	log.Printf("[INFO] env.BaseURLString: %s, env.BaseURL.Path: %s", env.BaseURLString, env.BaseURL.Path)
+
 	model.Initialize(env.BaseURL.Path)
-	ap := api.NewAPI().
-		BaseURL(*env.BaseURL).
+	ap, err := api.NewAPI()
+	if err != nil {
+		log.Fatalf("Error calling NewAPI: %v", err)
+	}
+	ap = ap.BaseURL(*env.BaseURL).
```
[the remark this answers is not in the document]
Comment (jancona) [reply server/server.go:80]: Good catch. I think I used it at some point, but not now. I removed it.

File: api/api.go [hunk 4]
```diff
@@ -46,12 +58,18 @@ type PubSubConfig struct {
...
 // NewAPI builds an API
-func NewAPI() *API {
+func NewAPI() (*API, error) {
 	api := &API{
 		baseURL: url.URL{},
 	}
 	api.Validate(validator.New())
-	return api
+	var err error
+	api.userCache, err = lru.New2Q(100)
```
[the remark this answers is not in the document]
Comment (jancona) [reply api/api.go:63]: Here's the discussion that led me to the current model: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

**Pull request merged**